toolchain/check/dataflow_analysis.cpp

```diff
@@ -0,0 +1,345 @@
+      CollectEntityNamesFromPattern(sem_ir, var_storage->pattern_id, names);
+      for (auto [entity_id, def_id] : names) {
+        infos.push_back({entity_id, def_id});
+      }
+    }
+  } else if (auto ref_bind = inst.TryAs<SemIR::RefBinding>()) {
+    infos.push_back({ref_bind->entity_name_id, inst_id});
+  } else if (auto val_bind = inst.TryAs<SemIR::ValueBinding>()) {
+    infos.push_back({val_bind->entity_name_id, inst_id});
+  } else if (auto name_ref = inst.TryAs<SemIR::NameRef>()) {
+    // NameRef.value_id points to the binding (RefBinding/ValueBinding).
+    auto binding_id = name_ref->value_id;
+    auto binding_inst = sem_ir.insts().Get(binding_id);
+    if (auto ref_bind = binding_inst.TryAs<SemIR::RefBinding>()) {
+      infos.push_back({ref_bind->entity_name_id, binding_id});
+    } else if (auto val_bind = binding_inst.TryAs<SemIR::ValueBinding>()) {
+      infos.push_back({val_bind->entity_name_id, binding_id});
+    }
+  }
+  return infos;
+}
+
+// This builds facts needed for carrying out dataflow analysis.
+static auto BuildDataflowFacts(const SemIR::File& sem_ir,
+                               SemIR::FunctionId function_id) -> DataflowFacts {
+  DataflowFacts facts;
+  const auto& function = sem_ir.functions().Get(function_id);
+
+  if (function.body_block_ids.empty()) {
+    return facts;
+  }
+
+  // Track ref params to treat assignments as uses.
+  Set<int32_t> ref_params;
+
+  // Collect definitions from parameters.
+  if (function.param_patterns_id.has_value()) {
+    auto param_patterns = sem_ir.inst_blocks().Get(function.param_patterns_id);
+    for (auto pattern_id : param_patterns) {
+      llvm::SmallVector<std::pair<SemIR::EntityNameId, SemIR::InstId>>
+          entity_names;
+      CollectEntityNamesFromPattern(sem_ir, pattern_id, entity_names);
+      for (auto [entity_name_id, def_inst_id] : entity_names) {
+        // Use the pattern_id as the instruction ID for the definition.
+        facts.defs.Insert(Fact{def_inst_id.index, entity_name_id.index});
+
+        // Identify ref parameters.
+        auto inst = sem_ir.insts().Get(pattern_id);
+        if (inst.Is<SemIR::RefParamPattern>()) {
+          ref_params.Insert(entity_name_id.index);
+        }
+      }
+    }
+  }
+
+  for (const auto& block_id : function.body_block_ids) {
+    const auto& block = sem_ir.inst_blocks().Get(block_id);
+
+    // Emit leader fact for non-empty blocks.
+    if (!block.empty()) {
+      facts.leaders.Insert(Fact{block_id.index, block.front().index});
+    }
+
+    // First pass: identify LHS of assignments to avoid counting them as uses.
+    Set<SemIR::InstId> assigned_lhs;
+    for (const auto& inst_id : block) {
+      auto inst = sem_ir.insts().Get(inst_id);
+      if (auto assign = inst.TryAs<SemIR::Assign>()) {
+        assigned_lhs.Insert(assign->lhs_id);
+      }
+    }
+
+    for (size_t i = 0; i < block.size(); ++i) {
+      auto inst_id = block[i];
+      auto inst = sem_ir.insts().Get(inst_id);
```

Comment:
What made you choose not to do a range-based for loop here? Note llvm provides `llvm::enumerate` if you specifically needed an index